src/mds/FSMap.cc

```diff
@@ -0,0 +1,762 @@
+      __s32 s;
+      ::decode(s, p);
+      legacy_mds_map.cas_pool = s;
+    } else {
+      ::decode(legacy_mds_map.data_pools, p);
+      ::decode(legacy_mds_map.cas_pool, p);
+    }
+
+    // kclient ignores everything from here
+    __u16 ev = 1;
+    if (struct_v >= 2)
+      ::decode(ev, p);
+    if (ev >= 3)
+      ::decode(legacy_mds_map.compat, p);
+    else
+      legacy_mds_map.compat = get_mdsmap_compat_set_base();
+    if (ev < 5) {
+      __u32 n;
+      ::decode(n, p);
+      legacy_mds_map.metadata_pool = n;
+    } else {
+      ::decode(legacy_mds_map.metadata_pool, p);
+    }
+    ::decode(legacy_mds_map.tableserver, p);
+    ::decode(legacy_mds_map.in, p);
+    ::decode(legacy_mds_map.inc, p);
+    ::decode(legacy_mds_map.up, p);
+    ::decode(legacy_mds_map.failed, p);
+    ::decode(legacy_mds_map.stopped, p);
+    if (ev >= 4)
+      ::decode(legacy_mds_map.last_failure_osd_epoch, p);
+    if (ev >= 6) {
+      ::decode(legacy_mds_map.ever_allowed_snaps, p);
+      ::decode(legacy_mds_map.explicitly_allowed_snaps, p);
+    } else {
+      legacy_mds_map.ever_allowed_snaps = true;
+      legacy_mds_map.explicitly_allowed_snaps = false;
+    }
+    if (ev >= 7)
+      ::decode(legacy_mds_map.inline_data_enabled, p);
+
+    if (ev >= 8) {
+      assert(struct_v >= 5);
+      ::decode(enabled, p);
+      ::decode(legacy_mds_map.fs_name, p);
+    } else {
+      if (epoch > 1) {
+        // If an MDS has ever been started, epoch will be greater than 1,
+        // assume filesystem is enabled.
+        enabled = true;
+      } else {
+        // Upgrading from a cluster that never used an MDS, switch off
+        // filesystem until it's explicitly enabled.
+        enabled = false;
+      }
+    }
+
+    if (ev >= 9) {
+      ::decode(legacy_mds_map.damaged, p);
+    }
+    // We're upgrading, populate fs_list from the legacy fields
+    assert(filesystems.empty());
+    auto migrate_fs = std::make_shared<Filesystem>(); 
+
+    *migrate_fs = legacy_fs;
+    migrate_fs->fscid = FS_CLUSTER_ID_ANONYMOUS;
+    migrate_fs->mds_map.fs_name = "default";
+    legacy_client_fscid = migrate_fs->fscid;
+    compat = migrate_fs->mds_map.compat;
+    enable_multiple = false;
+  } else {
+    ::decode(epoch, p);
+    ::decode(next_filesystem_id, p);
+    ::decode(legacy_client_fscid, p);
+    ::decode(compat, p);
+    ::decode(enable_multiple, p);
+    std::vector<Filesystem> fs_list;
+    ::decode(fs_list, p);
+    filesystems.clear();
+    for (std::vector<Filesystem>::const_iterator fs = fs_list.begin(); fs != fs_list.end(); ++fs) {
+      filesystems[fs->fscid] = std::make_shared<Filesystem>(*fs);
+    }
+
+    ::decode(mds_roles, p);
+    ::decode(standby_daemons, p);
+    ::decode(standby_epochs, p);
+  }
+
+  DECODE_FINISH(p);
+}
+
+
+void Filesystem::encode(bufferlist& bl) const
+{
+  ENCODE_START(1, 1, bl);
+  ::encode(fscid, bl);
+  bufferlist mdsmap_bl;
+  mds_map.encode(mdsmap_bl, CEPH_FEATURE_PGID64 | CEPH_FEATURE_MDSENC);
```

Comment:
Another hard-coded feature set for encoding? Unless there's something specific about these we'll want to fix that in a follow-up as well.